### ISSUE TITLE
Question about "iIoU = nan", how to interpret this?

### ISSUE THREAD

**Issue by Almo1982** [mcordts/cityscapesScripts#10]
Hi again Marius,

everything works so far but another question came up i didnt find an answer for yet:

i trained some nets with data from your dataset. when i calculate IoU and iIoU then with your script evalPixelLevelSemanticLabeling.py i get for example the result like attached:
![iou](https://cloud.githubusercontent.com/assets/18659261/17701776/6c1e8aa6-63cb-11e6-8cf0-73c326d04337.png)

My question is, why do i get a "nan" for the result of iIoU of almost all classes? 

When i take a look in your script there are following classes defined for calculating iIoU:

args.avgClassSize       = {
    "bicycle"    :  4672.3249222261 ,
    "caravan"    : 36771.8241758242 ,
    "motorcycle" :  6298.7200839748 ,
    "rider"      :  3930.4788056518 ,
    "bus"        : 35732.1511111111 ,
    "train"      : 67583.7075812274 ,
    "car"        : 12794.0202738185 ,
    "person"     :  3462.4756337644 ,
    "truck"      : 27855.1264367816 ,
    "trailer"    : 16926.9763313609 ,
}

So i expect at least for these classes not a "nan" as result?! And why did it work for "car" and "person" in my case? (well, its 0 but at least it does something :-) )

For the IoU, the "nan" for the train class means that there wasnt even one train-class in my training data set, correct?

But i dont understand all the other "nan", please help me out!

Thank you!
Stefan

**Comment by Almo1982** [comment]
Hi,

ok, one step further again ;-)

i think i understand now that only the classes that are marked in labels.py with "hasInstances" are respected for calculating iIoU!

I did another training and got better results for IoU, see attached screenshot.

There are less "nan" in the iIoU column now, but they are all 0.000, why? how exactly is this calculated?

for class "rider" and "motocycle" i still have iIoU = nan, ok, might depend on the score of IoU = 0.000.

But e.g. for car what is detected pretty good with IoU = 0.836, why is the iIoU = 0.000 there???

thank you!

Stefan
![screenshot from 2016-08-17 19 48 16](https://cloud.githubusercontent.com/assets/18659261/17747268/53218794-64b4-11e6-80d3-02f4d628d415.png)

**Comment by mcordts** [comment]
Hi Stefan,

most likely you modified the ground truth data. Please just freshly
download the folder gtFine and leave it unchanged, no rescaling or anything.
The script needs access to the instance annotations. Also it only makes
sense if you evaluate on a sensible number of images, i.e. our complete
validation set.

Best,
Marius

2016-08-17 19:54 GMT+02:00 Almo1982 <notifications@github.com>:

**Comment by Almo1982** [comment]
Hi Marius,

thank you for your Response!

The script has access to the instanceIds-annotations, they are in the same folder than the labelIds-annotations! 
You are right, i resized the original ground truth Images "labelIds", "labelTrainIds" and "instanceIds" to size 512x256 and used them for Training (better siad i just used the labelTrainIds-images for Training) and then calculating IoU and iIoU with the labelIds-images and instanceIds-images of size 512x256. I used openCV with INTER_NEAREST transformatoin, this worked well,pixel vlaues are correct.

Do you think the Resolution causes the Problem? I still have the original data of the dataset so i will give it a try! But my prediction Images are of size 512x256, will this work then with gt Images of 2048x1024 or do i have to upscale my prediction Images to 2048x1024 then?

I use 233 Images (all Images of Folder "lindau" and "munster", i skipped frankfurt) for Evaluation, isnt that enough?

thanks!
Stefan

**Comment by mcordts** [comment]
Yes, you need to upscale your predictions.
233 image might be enough, at least if the prior is >0 for all labels of
interest.

Best,
Marius

2016-08-18 8:47 GMT+02:00 Almo1982 <notifications@github.com>:

**Comment by Almo1982** [comment]
ok, but don't i loose Information then when upscaling with factor 4?

Should i do this with INTER_NEAREST again or is there another (better) Transformation method in this case? Maybe you can give me a hint for the best upscale method before i am trying things out for hours again ;-)

Thank you!
Stefan

**Comment by mcordts** [comment]
The task of this benchmark is to predict a label per pixel. The images have
2048x1024 pixels. How to achieve this is part of the method, not of the
dataset.

If you read the paper carefully and study the current benchmark results,
you will notice that downscaling the input image at all is generally not
the best idea.

2016-08-18 9:13 GMT+02:00 Almo1982 <notifications@github.com>:

**Comment by Almo1982** [comment]
Hi Marius,

the task of the Benchmark is clear for me i think, i only have an understanding Problem why it should work with 2048x1024 and not with 512x256...

If ALL Images have 512x256 Pixels, where is the difference (except of the less amount of Pixels of course)? i can also predict a Label per Pixel in this case?! 

It Looks like the calculation of IoU (comparing the prediction Images with the labelIds-annotations) is working fine on base of the 512x256 Images...so what is not clear to me why it doesnt work for the instanceIds-annotations then that have same size?!

But i will give it a try with the original size and upscaling my prediction Images...

Thank you again for your comments, i will take a deeper look into the paper again, maybe this helps for my understanding... 

Regards
Stefan

**Comment by mcordts** [comment]
If you want to submit to the evaluation server, you need to have full resolution anyways. 

The eval script does work for a changed resolution, but I am pretty sure that you messed up with the instance IDs while downscaling. Since I don't want to debug you code, just upscale your results and use the original data.

Best,
Marius

**Comment by Almo1982** [comment]
Hi Marius,

ok, understand that. for now i dont want to use my results for your Evaluation Server. Maybe i will do this later on with the upscaled, full Resolution.

I didnt have the Chance to test it again, will do this on the Weekend for sure! I was checking the scripts again, i think i have to divide the avgClassSize Parameters in evalPixelLevelSemanticLabeling.py by 16 because of my downscaling of the Image width and Image height of factor 4. then i have the same Ratio for the average class sizes like with full Resolution. Maybe this was the error in the calculation process?! Would make sense to me so far...

I will give it a try on the Weekend. But this should be the same like evaluating with the Standard Parameter values and full Resolution then?!

I checked downscaled instanceIds-images again, Pixel values should be correct, i dont think These Images are messed up...but i will also give it a try with full Resolution by upscalinf my prediction Images.

Thank you very much!

Regards
Stefan

**Comment by mcordts** [comment]
The instance ID images have 16 bits per pixel, which is not supported by
all png libraries. Cars have pixel values of 26000, 26001, ... . Better
double check that.

2016-08-19 8:41 GMT+02:00 Almo1982 <notifications@github.com>:

**Comment by Almo1982** [comment]
Ah,ok, that makes a difference then, true! Thank you for that, i will double check this!

but i still have to divide avgClassSize values by 16 because of my downscaling factor of 4, correct? 

Regards,
Stefan

**Comment by mcordts** [comment]
Yes, dividing by a factor of 16 should be correct.

2016-08-19 10:35 GMT+02:00 Almo1982 <notifications@github.com>:

**Comment by Almo1982** [comment]
Hi Marius,

you were absolutely right! during downsampling the instanceIds-Images they became uint8! I downsampled them again with respect to the dtype and they are resolution 512x256 now with dtype "uint16". I also divided the avgClassSize-Values in evalPixelLevelSemanticLabeling.py by 16 and get the following results now like attached. This makes much more sense to me and seems to be correct now!

Thank you again for your great support!

Regards
Stefan

![screenshot from 2016-08-19 23 40 32](https://cloud.githubusercontent.com/assets/18659261/17825195/e5ece8e4-6666-11e6-87e7-46f0b51492d3.png)
